Detectors/EMCAL/reconstruction/src/CaloRawFitterGamma2.cxx

```diff
@@ -0,0 +1,172 @@
+}
+
+CaloFitResults CaloRawFitterGamma2::evaluate(const std::vector<Bunch>& bunchlist,
+                                             std::optional<unsigned int> altrocfg1, std::optional<unsigned int> altrocfg2)
+{
+  float time = 0;
+  float amp = 0;
+  float chi2 = 0;
+  int ndf = 0;
+  bool fitDone = false;
+
+  auto [nsamples, bunchIndex, ampEstimate,
+        maxADC, timeEstimate, pedEstimate, first, last] = preFitEvaluateSamples(bunchlist, altrocfg1, altrocfg2, mAmpCut);
+
+  if (ampEstimate >= mAmpCut) {
+    time = timeEstimate;
+    int timebinOffset = bunchlist.at(bunchIndex).getStartTime() - (bunchlist.at(bunchIndex).getBunchLength() - 1);
+    amp = ampEstimate;
+
+    if (nsamples > 2 && maxADC < constants::OVERFLOWCUT) {
+      doParabolaFit(timeEstimate - 1, amp, time);
+      mNiter = 0;
+      fitDone = doFit_1peak(first, nsamples, amp, time, chi2);
+
+      if (!fitDone) {
+        amp = ampEstimate;
+        time = timeEstimate;
+        chi2 = 1.e9;
+      }
+
+      time += timebinOffset;
+      timeEstimate += timebinOffset;
+      ndf = nsamples - 2;
+    }
+  }
+
+  if (fitDone) {
+    float ampAsymm = (amp - ampEstimate) / (amp + ampEstimate);
+    float timeDiff = time - timeEstimate;
+
+    if ((TMath::Abs(ampAsymm) > 0.1) || (TMath::Abs(timeDiff) > 2)) {
+      amp = ampEstimate;
+      time = timeEstimate;
+      fitDone = false;
+    }
+  }
+  if (amp >= mAmpCut) {
+    if (!fitDone) {
+      std::default_random_engine generator;
+      std::uniform_real_distribution<float> distribution(0.0, 1.0);
+      amp += (0.5 - distribution(generator));
+    }
+    time = time * constants::EMCAL_TIMESAMPLE;
+    time -= mL1Phase;
+
+    return CaloFitResults(maxADC, pedEstimate, mAlgo, amp, time, (int)time, chi2, ndf);
+  }
+  return CaloFitResults(-1, -1);
+}
+
+bool CaloRawFitterGamma2::doFit_1peak(int ifirst, int nS, float& A0, float& t0, float& chi2)
+{
+  // fit using gamma-2 function 	(ORDER =2 assumed)
+
+  if (nS < 3)
+    return false;
+  if (mNiter > mNiterationsMax)
+    return false;
+
+  double expi, ti, g_i, g_1i, g_2i, g1_1, g2_i, gp_i, g2p_i, q1_i, q2_i, D, dA, dt, delta;
+  double c11 = 0;
+  double c12 = 0;
+  double c21 = 0;
+  double c22 = 0;
+  double d1 = 0;
+  double d2 = 0;
+  chi2 = 0;
+
+  mNiter++;
+
+  for (int i = 0; i < nS; i++) {
+
+    ti = (i - t0) / constants::TAU;
+    if ((ti + 1) < 0)
+      continue;
+
+    expi = TMath::Exp(-2 * ti);
+    g_2i = expi;
+    g_1i = (ti + 1) * g_2i;
+    g_i = (ti + 1) * g_1i;
+    gp_i = 2 * (g_i - g_1i);
+    g2p_i = 2 * g_i * gp_i;
+    q1_i = g_2i * (2 * ti + 1);
+    q2_i = g_1i * g_1i * (4 * ti + 1);
+    c11 += (getReversed(i) - A0 * 2 * g_i) * gp_i;
+    c12 += g_i * g_i;
+    c21 += getReversed(i) * q1_i - A0 * q2_i;
+    c22 += g_i * g_1i;
+    delta = A0 * g_i - getReversed(i);
+    d1 += delta * g_i;
+    d2 += delta * g_1i;
+    chi2 += (delta * delta);
+  }
+
+  D = c11 * c22 - c12 * c21;
+
+  if (TMath::Abs(D) < 1.e-9)
+    return false;
```

Comment:
I assume this is a check fro 0. In this case the comparison should be done against DBL_EPSILON which is provided by the standard library on the target system. In case of D=0 an exception (FitException - to be implemented as nested class) should be thrown rather than returning a boolean.

Reply:
Still need to do the exception.

Reply:
Yes, in case the comparison is 0 this is where the exception is thrown.